### ISSUE TITLE
Adding TensorFlow Hub KerasLayer to Sequential Model Raises ValueError

### ISSUE THREAD

**Issue by ruddyscent** [tensorflow/tensorflow#63849]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.16.1

### Custom code

No

### OS platform and distribution

Ubuntu 22.04.3 LTS

### Python version

Python 3.11.0rc1

### CUDA/cuDNN version

12.3.0

### Current behavior?

I can execute the following code without any issues in TensorFlow 2.15.0 and TensorFlow Hub 1.16.1. However, when I upgrade the TensorFlow version to 2.16.0 or above, I encounter an error stating that `KerasLayer` cannot be added to the Sequential model.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import tensorflow_hub as hub

image_size = 224
URL = "https://tfhub.dev/tensorflow/efficientnet/b0/feature-vector/1"

model = tf.keras.Sequential([
        hub.KerasLayer(URL, input_shape=(image_size, image_size, 3))
])
```

### Relevant log output

```shell
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[29], line 1
----> 1 model = tf.keras.Sequential([
      2         feature_extractor,
      3         tf.keras.layers.Dense(2, activation = 'softmax')
      4 ])
      6 model.build([None, image_size, image_size, 3])
      7 model.summary()

File /usr/local/lib/python3.10/dist-packages/keras/src/models/sequential.py:70, in Sequential.__init__(self, layers, trainable, name)
     68 if layers:
     69     for layer in layers:
---> 70         self.add(layer, rebuild=False)
     71     self._maybe_rebuild()

File /usr/local/lib/python3.10/dist-packages/keras/src/models/sequential.py:92, in Sequential.add(self, layer, rebuild)
     90         layer = origin_layer
     91 if not isinstance(layer, Layer):
---> 92     raise ValueError(
     93         "Only instances of `keras.Layer` can be "
     94         f"added to a Sequential model. Received: {layer} "
     95         f"(of type {type(layer)})"
     96     )
     97 if not self._is_layer_name_unique(layer):
...
    101         "the name of a layer in this model. Update the `name` argument "
    102         "to pass a unique name."
    103     )

ValueError: Only instances of `keras.Layer` can be added to a Sequential model. Received: <tensorflow_hub.keras_layer.KerasLayer object at 0x7a4ac7e30f40> (of type <class 'tensorflow_hub.keras_layer.KerasLayer'>)
```

**Comment by Aloqeely** [comment]
After investigating the code, I found a potential cause of the issue in `tensorflow_hub/keras_layer.py` lines 26-31:
```python
# Use Keras 2.
version_fn = getattr(tf.keras, "version", None)
if version_fn and version_fn().startswith("3."):
  import tf_keras as keras
else:
  keras = tf.keras
```
I believe this conditional import is causing the `isinstance(layer, Layer)` check in `Sequential.add` to return `False` in certain cases.

**Comment by SuryanarayanaY** [comment]
Hi @ruddyscent ,

TFHub has dependency on `tf_keras` package (i.e Keras2) as per the setup.py of TFHub.

https://github.com/tensorflow/hub/blob/ff72e25fef44bd67d5c14fb5328aa44e303e3404/tensorflow_hub/pip_package/setup.py#L31

Since TF2.16 comes with Keras3 the problem arises. As a workaround, you can install tf_keras package and set environment variable TF_USE_LEGACY_KERAS=1 to ensure Keras2 will be used with tf.keras.

**Comment by SuryanarayanaY** [comment]
> After investigating the code, I found a potential cause of the issue in `tensorflow_hub/keras_layer.py` lines 26-31:
> 
> ```python
> # Use Keras 2.
> version_fn = getattr(tf.keras, "version", None)
> if version_fn and version_fn().startswith("3."):
>   import tf_keras as keras
> else:
>   keras = tf.keras
> ```
> 
> Depending on the Keras version, the module might either import `tf_keras` or directly use `tf.keras`, the latter causes the `isinstance(layer, Layer)` check in `Sequential.add` to return `False` for `hub.KerasLayer`, even though it inherits from `keras.layers.Layer`

Hi, When I tried with TF2.16v on Colab environment the error stack seems generated from the _ensure_keras_2_importable() function from `tensorflow_hub/__init__.py` as per attached [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/5585bf7974198ca45bacdb9e84db91d4/63849_tf2-16.ipynb).

**Comment by SuryanarayanaY** [comment]
> ```python
> # Use Keras 2.
> version_fn = getattr(tf.keras, "version", None)
> if version_fn and version_fn().startswith("3."):
>   import tf_keras as keras
> else:
>   keras = tf.keras
> ```

Even this code seems confusing to me. If version_fn results in Keras3 then it tries to import tf_keras to use Keras2 else it assumes Keras2 is alreday there(i.e TF<=2.15v) hence it takes keras=tf.keras. 

But `version_fn = getattr(tf.keras, "version", None)` returns None making this to use tf.keras (i.e Keras3) which ic not compatible with TF_Hub.

**Comment by SuryanarayanaY** [comment]
> ```python
> # Use Keras 2.
> version_fn = getattr(tf.keras, "version", None)
> if version_fn and version_fn().startswith("3."):
>   import tf_keras as keras
> ```

From the code above if version_fn is not None and if it starts with "3" (which means Keras 3 found) in that case it is importing tf_keras package as keras. Please note that tf_keras package is for Keras2. This indicates TFHub supports only Keras2 package.

The else part it assumes that version_fn doesn't starts with "3" which case it assumes Keras2 installed with TF package and it marks keras=tf.keras. But for any case if version_fn becomes None and if Keras3 installed with TF then tf.keras will become Keras3 which might a problem. 

@Aloqeely , Could you please import TF2.16 and confirm what will be the version_fn output ?

**Comment by Aloqeely** [comment]
Here you go
```python
import tensorflow as tf

version_fn = getattr(tf.keras, "version", None)
print("TF Version: " + tf.__version__)
print("TF Keras Version: " + version_fn())
```
Output:
```
TF Version: 2.16.1
TF Keras Version: 3.0.5
```

**Comment by Aloqeely** [comment]
> Hi @ruddyscent ,
> 
> TFHub has dependency on `tf_keras` package (i.e Keras2) as per the setup.py of TFHub.
> 
> https://github.com/tensorflow/hub/blob/ff72e25fef44bd67d5c14fb5328aa44e303e3404/tensorflow_hub/pip_package/setup.py#L31
> 
> Since TF2.16 comes with Keras3 the problem arises. As a workaround, you can install tf_keras package and set environment variable TF_USE_LEGACY_KERAS=1 to ensure Keras2 will be used with tf.keras.

If this workaround fixes ruddyscent's problem, then I think we should mark this issue as resolved, because the problem is that tensorflow hub does not support Keras 3, so it is not relevant to this repository.

**Comment by SuryanarayanaY** [comment]
> 3.0.5

This means when tf.keras version is 3.x then Hub suggesting to import tf_keras as keras which is a Keras2 package. For that we should install with tf_keras package using `pip install tf_keras`. But from the error log the package is having keras/src which seems to be from Keras3 package for me.

**Comment by Aloqeely** [comment]
> This means when tf.keras version is 3.x then Hub suggesting to import tf_keras as keras which is a Keras2 package. For that we should install with tf_keras package using `pip install tf_keras`. But from the error log the package is having keras/src which seems to be from Keras3 package for me.

Yep, hub does that.
The error log is caused by `tf.keras.Sequential` (Keras 3, hence the error is produced from keras/src) after it received a `hub.KerasLayer` (Keras 2)

**Comment by sj-yoon** [comment]
win10, no gpu, no colab

tensorflow_hub를 사용하시려면 다음과 같이 해보세요
import tensorflow as tf
import tensorflow_hub as hub
import tf_keras as keras  """tensorflow_hub 설치시 자동으로 설치됩니다."

...
...
model = keras.Sequential([
      hub,KerasLayer(url, input_shape=(size, size, 3)
...
...
imgfile = keras.utils.get_file('image.jpg', 'https://..........')
작동이 잘됩니다.

**Comment by ruddyscent** [comment]
Thank you, @sj-yoon, for your assistance. 

The solution provided by @SuryanarayanaY works well for my situation.

**Comment by alishhde** [comment]
I had the same issue, and I could fix it with the solution provided by @SuryanarayanaY. Thanks 

Others can check my notebook for steps to fix it at this [page](https://www.kaggle.com/code/alishohadaee/dl-transfer-learning-on-mobilenet-v2).

**Comment by tilakrayal** [comment]
Could you please check whether this issue is resolved with the latest tensorflow v2.17 which contains Keras 3.0 version. As Keras3 now built to support multiple backends there are some changes in design.

Also this issue is more related to Keras, please try to raise the issue in keras-team/Keras repo for the quick resolution. Thank you!

**Comment by Dobiasd** [comment]
> Could you please check whether this issue is resolved with the latest tensorflow v2.17 which contains Keras 3.0 version.

The problem still exists with the latest `tensorflow` version (currently `2.17.0`) as the following minimal `Dockerfile` shows:

```Dockerfile
FROM python:3.12.5

RUN pip3 install tensorflow==2.17.0 tensorflow-hub==0.16.1

RUN echo 'import tensorflow as tf' > main.py
RUN echo 'import tensorflow_hub as hub' >> main.py
RUN echo 'model = tf.keras.Sequential([hub.KerasLayer("https://tfhub.dev/tensorflow/efficientnet/b0/feature-vector/1", input_shape=(224, 224, 3))])' >> main.py

RUN python main.py
```

```bash
docker build --rm --progress=plain .
```

Output:
```
...
6.058   File "//main.py", line 3, in <module>
6.058     model = tf.keras.Sequential([hub.KerasLayer("https://tfhub.dev/tensorflow/efficientnet/b0/feature-vector/1", input_shape=(224, 224, 3))])
6.058             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
6.058   File "/usr/local/lib/python3.12/site-packages/keras/src/models/sequential.py", line 74, in __init__
6.058     self.add(layer, rebuild=False)
6.058   File "/usr/local/lib/python3.12/site-packages/keras/src/models/sequential.py", line 96, in add
6.058     raise ValueError(
6.058 ValueError: Only instances of `keras.Layer` can be added to a Sequential model. Received: <tensorflow_hub.keras_layer.KerasLayer object at 0x7ff1f3d0be60> (of type <class 'tensorflow_hub.keras_layer.KerasLayer'>)
```

**Comment by sondosaabed** [comment]
Was this fixed for tf 2.16 !

**Comment by sondosaabed** [comment]
I had to use a lower version of tensorflow, this has to be bug!

**Comment by rlcauvin** [comment]
I agree that `hub.KerasLayer` should be fixed to work with Keras 3.

Meanwhile, however, Keras 3 includes `keras.layers.TFSMLayer` for loading TensorFlow models as Keras layers. Here is an example of loading a TensorFlow Hub model as a Keras layer:
```
path = hub.resolve(handle="https://tfhub.dev/google/universal-sentence-encoder/4")
use_layer = keras.layers.TFSMLayer(
      filepath=path,
      call_endpoint="serving_default",
      trainable=True,
      name="use_layer",
      dtype="string"
    )
```
Unfortunately, at this time, there is a [bug](https://github.com/keras-team/keras/pull/20995) in `keras.layers.TFSMLayer` where it doesn't properly initialize the variables upon loading. The bug is fixed in the master repo but not in the current Keras 3.9 release. If you want to try it before the next release, you may install the Keras master with:
```
!pip install git+https://github.com/keras-team/keras.git@master
```

**Comment by Anand-Velpuri** [comment]
Hi @ruddyscent , @alishhde , @SuryanarayanaY,

Thanks for reporting this issue.

I believe the following code snippet provides a solution or a starting point to address the problem. It outlines how to properly build a Keras model that incorporates a TensorFlow Hub layer.
```python
def hub_layer_fn(x):
    return hub.KerasLayer(MODEL_URL, trainable=False)(x)
```

```python
import tensorflow as tf
import tensorflow_hub as hub

# Create a function which builds a Keras model
def create_model(input_shape=INPUT_SHAPE, output_shape=OUTPUT_SHAPE, model_url=MODEL_URL):
    print("Building model with:", model_url)

    # Setup the model layers
    model = tf.keras.Sequential([
        tf.keras.layers.Lambda(hub_layer_fn, name="hub_lambda"),  # Wrap the hub.KerasLayer in Lambda
        tf.keras.layers.Dense(units=output_shape, activation="softmax")  # Output layer with softmax activation
    ])

    # Compile the model
    model.compile(
        loss=tf.keras.losses.CategoricalCrossentropy(),
        optimizer=tf.keras.optimizers.Adam(),
        metrics=["accuracy"]
    )

    # Build the model
    model.build((None, *input_shape))  # Ensure input shape is unpacked correctly

    return model
```

This function `create_model` takes `input_shape`, `output_shape`, and `model_url` as parameters and constructs a `tf.keras.Sequential` model. It uses a `tf.keras.layers.Lambda` layer to wrap the TensorFlow Hub layer (represented by `hub_layer_fn`, which you would need to define based on how you are loading the Hub module). This approach can sometimes resolve compatibility or layering issues.

Please ensure you have `INPUT_SHAPE`, `OUTPUT_SHAPE`, `MODEL_URL`, and `hub_layer_fn` defined according to your specific use case.

Another Example:
```python
resnet_url = "https://www.kaggle.com/models/google/resnet-v2/TensorFlow2/50-feature-vector/2"

efficientnet_url = "https://www.kaggle.com/models/tensorflow/efficientnet/TensorFlow2/b0-feature-vector/1"
```
```python
# Import dependencies
import tensorflow as tf
import tensorflow_hub as hub
from tensorflow.keras import layers
```

```python
def create_model(model_url, num_classes=10):
  """
  Takes a TensorFlow Hub URL and creates a Keras Sequential model with it.

  Args:
    model_url (str): A TensorFlow Hub feature extraction URL.
    num_classes (int): Number of output neurons in the output layer,
    should be equal to number of target classes, default 10.
  
  Returns:
    An uncompiled keras Sequential model with model_url as feature extarctor
    layer and Dense output layer with num_classes output neurons.
  """
  # Download the pretrained model and save it as a keras layer
  feature_extractor_layer = hub.KerasLayer(model_url,
                                           trainable=False, # Freeze the already learned patterns
                                           name="feature_extraction_layer",
                                           input_shape=IMAGE_SHAPE+(3,)
                                           )
  
  # Create our own model
  model = tf.keras.Sequential([
      layers.Lambda(feature_extractor_layer, name="hub_lambda"),
      layers.Dense(num_classes, activation="softmax", name="output_layer")
  ])

  model.build((None, *(IMAGE_SHAPE+(3,))))

  return model
```

```python
# Create Resnet model
resnet_model = create_model(resnet_url,
                            num_classes=train_data_10_percent.num_classes
                            )
```

```python
resnet_model.summary()
```

Simply put this in your notebook's top cell if you don't want to do any of these tasks:
```python
import os
os.environ["TF_USE_LEGACY_KERAS"] = "1"
```

For more info check out my Dog Vision Project Notebook: https://github.com/Anand-Velpuri/Dog-Vision/blob/main/dog-vision.ipynb

Let me know if this helps resolve the issue!

Best regards,
@Anand-Velpuri

**Comment by AKuinkel** [comment]
If we take a look @ our sequential.py here, we can see why the error is raised:

<img width="554" height="244" alt="Image" src="https://github.com/user-attachments/assets/60d20f4b-7e62-48f3-b6af-001930129d03" />

Our isinstance is returning false even though it gets this from keras.layers.Layers

I read about this and apparently it's due to some compatibility issues between versions of  keras and tf casuing the error:


I didn't downgrade my tf version, or import tf_keras as keras (you can try these, since it has worked for people), wrapping it in lambda like this resolved the problem for me (since it's now proper tf.keras.layers.Layer, our isinstance returns true):

**CODE WITH ERROR:**
```python
model = tf.keras.Sequential([
 feature_extractor_layer
  tf.keras.layers.Dense(num_classes)
])
``` 

**Error resolved using lambda:**
```python
model = tf.keras.Sequential([
  tf.keras.layers.InputLayer(input_shape=(img_height,img_width, 3)),
  tf.keras.layers.Lambda(lambda x: feature_extractor_layer(x)), 
  tf.keras.layers.Dense(num_classes)
])
``` 

Here is a good link I found for this: https://blog.oofdev.com/2024/03/adding-tensorflow-hub-keraslayer-sequentail-raises-valueerror.html

**Comment by AKuinkel** [comment]
> If we take a look @ our sequential.py here, we can see why the error is raised:
> 
> <img alt="Image" width="554" height="244" src="https://private-user-images.githubusercontent.com/108271619/495959158-60d20f4b-7e62-48f3-b6af-001930129d03.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTkzMDMxNjQsIm5iZiI6MTc1OTMwMjg2NCwicGF0aCI6Ii8xMDgyNzE2MTkvNDk1OTU5MTU4LTYwZDIwZjRiLTdlNjItNDhmMy1iNmFmLTAwMTkzMDEyOWQwMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAwMVQwNzE0MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYjgzODlmYzk1YjhhMzY3OWE3ZGIxOGFlYmUyOTAwNWY1MjU3ZWRmODdlNDg2YjIzYTA4NjRiZDNjODRlY2NiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ToChXp4wM3sanrsWywK4W_QlIYT_sfNjvRb-nMs5qKE">
> Our isinstance is returning false even though it gets this from keras.layers.Layers
> 
> I read about this and apparently it's due to some compatibility issues between versions of keras and tf casuing the error:
> 
> I didn't downgrade my tf version, or import tf_keras as keras (you can try these, since it has worked for people), wrapping it in lambda like this resolved the problem for me (since it's now proper tf.keras.layers.Layer, our isinstance returns true):
> 
> **CODE WITH ERROR:**
> 
> model = tf.keras.Sequential([
>  feature_extractor_layer
>   tf.keras.layers.Dense(num_classes)
> ])
> **Error resolved using lambda:**
> 
> model = tf.keras.Sequential([
>   tf.keras.layers.InputLayer(input_shape=(img_height,img_width, 3)),
>   tf.keras.layers.Lambda(lambda x: feature_extractor_layer(x)), 
>   tf.keras.layers.Dense(num_classes)
> ])
> Here is a good link I found for this: https://blog.oofdev.com/2024/03/adding-tensorflow-hub-keraslayer-sequentail-raises-valueerror.html

**Update**: This lamda caused problems when trying to test my model, so I searched for another way and came accross this:

```python
version_fn = getattr(tf.keras, "version", None)
if version_fn and version_fn().startswith("3."):
import tf_keras as keras
else:
keras = tf.keras
```

import that into your file and then just change all your tf.keras (if that's what u have) to just keras.

Reference: https://github.com/tensorflow/tensorflow/issues/65419#issuecomment-2048298400

**Comment by mihaimaruseac** [comment]
This likely belongs in tensorflow/hub
